docs/api-reference/types/aggregate-root.md

```diff
@@ -0,0 +1,235 @@
+
+Initializes a new instance of the `AggregateRoot` class with the specified ID and restores it from the provided events.
+
+```csharp
+protected AggregateRoot(Guid id, IEnumerable<object> events);
+```
+
+**Parameters**:
+- `id` (`System.Guid`): The unique identifier for the aggregate.
+- `events` (`System.Collections.Generic.IEnumerable<object>`): The events to restore from.
+
+## Properties
+
+### Id
+
+Gets the unique identifier for this aggregate.
+
+```csharp
+public Guid Id { get; }
+```
+
+**Property Type**: `System.Guid`  
+**Accessibility**: `get`
+
+### ExpectedVersion
+
+Gets or sets the expected version this aggregate is at. This is used for optimistic concurrency control.
+
+```csharp
+public long ExpectedVersion { get; set; }
+```
+
+**Property Type**: `System.Int64`  
+**Accessibility**: `get`, `set`
+
+## Methods
+
+### RaiseEvent
+
+Raises an event, which will be recorded and applied to the aggregate.
+
+```csharp
+protected void RaiseEvent(object @event);
+```
+
+**Parameters**:
+- `event` (`System.Object`): The event to raise.
+
+**Remarks**: This method records the event and applies it to the aggregate by calling the appropriate `Apply` method.
+
+### RestoreFromEvents
+
+Restores this aggregate from the history of events.
+
+```csharp
+public void RestoreFromEvents(IEnumerable<object> events);
+```
+
+**Parameters**:
+- `events` (`System.Collections.Generic.IEnumerable<object>`): The events to restore from.
+
+**Exceptions**:
+- `System.ArgumentNullException`: Thrown when `events` is `null`.
+
+**Remarks**: This method applies each event in sequence to rebuild the aggregate's state.
+
+### UpdateWithEvents
+
+Updates this aggregate with the provided events, starting from the expected version.
+
+```csharp
+public void UpdateWithEvents(IEnumerable<object> events, long expectedVersion);
+```
+
+**Parameters**:
+- `events` (`System.Collections.Generic.IEnumerable<object>`): The events to update with.
+- `expectedVersion` (`System.Int64`): The expected version to start from.
+
+**Exceptions**:
+- `System.ArgumentNullException`: Thrown when `events` is `null`.
+- `System.InvalidOperationException`: Thrown when this aggregate does not have historical events or expected version mismatch.
+
+**Remarks**: This method checks that the expected version matches the aggregate's current version, and then applies each event in sequence.
+
+### TakeEvents
+
+Takes the recorded history of events from this aggregate.
+
+```csharp
+public object[] TakeEvents();
+```
+
+**Returns**: `System.Object[]` - The recorded events.
+
+**Remarks**: This method returns the recorded events and clears the aggregate's record of those events.
+
+## Usage
+
+The `AggregateRoot` class is designed to be subclassed by domain aggregates. Subclasses should:
+
+1. Define private `Apply` methods for each event type
+2. Use the `RaiseEvent` method to record and apply events
+3. Define public methods that represent domain operations
```

Comment:
This isn't quite right. The `Apply` methods need to be registered as handlers for those event types or they won't get used when rehydrating an aggregate from an event stream. This is typically done by calling `Register<T>(Apply)` for each event type in the constructor, or in a method that can be called from each constructor.

The same issue appears in multiple examples throughout.